[meson|scripts] update scripts

Neumann-A opened the pull request:
use native/cross files for meson and make crossbuilds on windows work

depends 

- [x] https://github.com/mesonbuild/meson/issues/7977

Neumann-A commented:
seems like a merged the wrong branch here once. The expat changes shouldn't be in here.

Neumann-A commented:
Current error:
```
Traceback (most recent call last):
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\mesonmain.py", line 140, in run
    return options.run_func(options)
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\msetup.py", line 253, in run
    app.generate()
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\msetup.py", line 161, in generate
    self._generate(env)
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\msetup.py", line 223, in _generate
    intr.backend.generate()
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\backend\ninjabackend.py", line 533, in generate
    self.write_rules(outfile)
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\backend\ninjabackend.py", line 1122, in write_rules
    b.count_rule_references()
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\backend\ninjabackend.py", line 342, in count_rule_references
    if self._should_use_rspfile():
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\backend\ninjabackend.py", line 336, in _should_use_rspfile
    return self.rule.length_estimate(infilenames,
  File "E:\vcpkg_cache\downloads\tools\meson\meson-0.56.0\mesonbuild\backend\ninjabackend.py", line 273, in length_estimate
    if chunk[1] == '{':
IndexError: string index out of range

```

waiting for upstream comment

Neumann-A committed:
Update meson to a recent commit to fix a bug with 0.56

Neumann-A committed:
rename static libraries built by meson by default on windows

Neumann-A committed:
apply cmake variables detection to meson.
use native/cross files
make arm builds work.

Neumann-A committed:
update munit

Neumann-A committed:
fix fribidi

Neumann-A committed:
fix _FLAGS variables

Neumann-A commented:
something is strange here:
without the  changes the following compiles: 
`[1/5] cc -Imunit_example.p -I. -I../src/3a51af77fd-8670ab3989.clean -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -std=c99 -fPIC -MD -MQ munit_example.p/example.c.o -MF munit_example.p/example.c.o.d -o munit_example.p/example.c.o -c ../src/3a51af77fd-8670ab3989.clean/example.c`
with these changes there is an error:
```
[1/5] /Library/Developer/CommandLineTools/usr/bin/cc -Imunit_example.p -I. -I../src/3a51af77fd-8670ab3989.clean -I/Users/vagrant/Data/installed/x64-osx/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -std=c99 -fPIC -O3 -DNDEBUG -MD -MQ munit_example.p/example.c.o -MF munit_example.p/example.c.o.d -o munit_example.p/example.c.o -c ../src/3a51af77fd-8670ab3989.clean/example.c
FAILED: munit_example.p/example.c.o 
/Library/Developer/CommandLineTools/usr/bin/cc -Imunit_example.p -I. -I../src/3a51af77fd-8670ab3989.clean -I/Users/vagrant/Data/installed/x64-osx/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -std=c99 -fPIC -O3 -DNDEBUG -MD -MQ munit_example.p/example.c.o -MF munit_example.p/example.c.o.d -o munit_example.p/example.c.o -c ../src/3a51af77fd-8670ab3989.clean/example.c
In file included from ../src/3a51af77fd-8670ab3989.clean/example.c:11:
../src/3a51af77fd-8670ab3989.clean/munit.h:29:10: fatal error: 'stdlib.h' file not found
#include <stdlib.h>
         ^~~~~~~~~~
1 error generated.
```
for me the cmd line looks virtually the same except that instead of only `cc` the full path to `/Library/Developer/CommandLineTools/usr/bin/cc` is used (and the extra compiler flags and includes). Can somebody with access to a OSX debug this? I don't understand what is happening here and the osx failure is the only thing blocking this PR. @JackBoosY @strega-nil @BillyONeal

JackBoosY commented:
@Neumann-A Trying...

JackBoosY commented:
Same error on osx machine by manually run that command:
```
vcpkg@vcpkg x64-osx-rel %  /Library/Developer/CommandLineTools/usr/bin/cc -Imunit_example.p -I. -I../src/3a51af77fd-8670ab3989.clean -I/Users/vcpkg/Documents/vcpkg/12945/vcpkg/installed/x64-osx/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -std=c99 -fPIC -O3 -DNDEBUG -MD -MQ munit_example.p/example.c.o -MF munit_example.p/example.c.o.d -o munit_example.p/example.c.o -c ../src/3a51af77fd-8670ab3989.clean/example.c
In file included from ../src/3a51af77fd-8670ab3989.clean/example.c:11:
../src/3a51af77fd-8670ab3989.clean/munit.h:29:10: fatal error: 'stdlib.h' file not found
#include <stdlib.h>
         ^~~~~~~~~~
1 error generated.
```

@strega-nil Can you please take a look?

Neumann-A committed:
set env SDKROOT and MACOSX_DEPLOYMENT_TARGET in configure_meson

Neumann-A committed:
need osx env while building and configuring

ras0219 commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -33,67 +33,388 @@
             set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/android.cmake")
         elseif(VCPKG_TARGET_IS_OSX)
             set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/osx.cmake")
         elseif(VVCPKG_TARGET_IS_FREEBSD)
```
I suspect this should be something besides `VVCPKG_TARGET_IS_FREEBSD`

Neumann-A replied:
fixed I still wished default toolchain would be in the triplets instead of `vcpkg_configure_`.

Neumann-A committed:
- remove unused env backup functions
- rename internal functions to be prefixed with _vcpkg_
- fix VCPKG_CHAINLOAD_TOOLCHAIN_FILE for ios/freebsd

Neumann-A committed:
replace _vcpkg with vcpkg since _ has special meaning in functions

strega-nil commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -33,67 +33,372 @@
-function(vcpkg_configure_meson)
-    # parse parameters such that semicolons in options arguments to COMMAND don't get erased
-    cmake_parse_arguments(PARSE_ARGV 0 _vcm "" "SOURCE_PATH" "OPTIONS;OPTIONS_DEBUG;OPTIONS_RELEASE")
+function(vcpkg_generate_native_file) #https://mesonbuild.com/Native-environments.html
```
```suggestion
function(vcpkg_internal_meson_generate_native_file) #https://mesonbuild.com/Native-environments.html
```

We want to note that these are _internal_, so you should be able to see if someone is using it (and could you make the same changes to all the other functions?)

strega-nil commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -33,67 +33,372 @@
+            message(FATAL_ERROR "Unsupported host architecture ${BUILD_ARCH}!" )
+        endif()
+    else() # TODO: add correct detection for OSX and Linux. Currently only x64 triplets are available in official vcpkg. 
+        set(BUILD_CPU_FAM x86_x64)
```
Note to self: I have to fix this soon yay

strega-nil commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -246,16 +246,19 @@ function(vcpkg_find_acquire_program VAR)
     set(HASH 22ea847d8017cd977664d0b13c889cfb13c89143212899a511be217345a4e243d4d8d4099700114a11d26a087e83eb1a3e2b03bdb5e0db48f10403184cd26619)
   elseif(VAR MATCHES "MESON")
-    set(MESON_VERSION 0.56.0)
+    set(MESON_VERSION 0.56.01)
```
Probably update tool-meson's version as well.

strega-nil reviewed (changes requested):
I'd like to see these changes made then I'll merge, thanks @Neumann-A :)

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -33,67 +33,388 @@
-function(vcpkg_configure_meson)
-    # parse parameters such that semicolons in options arguments to COMMAND don't get erased
-    cmake_parse_arguments(PARSE_ARGV 0 _vcm "" "SOURCE_PATH" "OPTIONS;OPTIONS_DEBUG;OPTIONS_RELEASE")
+function(add_to_env envvar value)
```
already removed since unused

Neumann-A committed:
rename functions to vcpkg_internal_meson_generate

Neumann-A committed:
update control tool-meson

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -33,67 +33,388 @@
             set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/osx.cmake")
         elseif(VVCPKG_TARGET_IS_FREEBSD)
+            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/ios.cmake")
+        elseif(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
```
already fixed

strega-nil commented:
Thanks so much @Neumann-A ! You've put a ton of work into making this a better experience and we really appreciate it 😄

c72578 commented:
Just for the records:
The meson version used in vcpkg is now at this commit of the master branch:
https://github.com/mesonbuild/meson/commit/91876b40316962620c1705ae14075ab46f8dd644

Neumann-A commented:
@c72578 yeah I had to restore to using some fixed commit of meson/master since meson seems to be horribly unstable and I did not want to wait for 0.61 for this to be merged .... I also considered to somehow add patches to `vcpkg_find_acquire_program` to patch the mess meson sometimes is.....
